src/bam/simulate_reads.rs

```diff
@@ -0,0 +1,175 @@
+use anyhow::Result;
+use bio::io::fasta;
+use rand::prelude::ThreadRng;
+use rand::Rng;
+use rust_htslib::bam;
+use rust_htslib::bam::Read;
+use std::path::Path;
+use uuid::Uuid;
+
+pub fn simulate_reads<P: AsRef<Path>>(
+    bam: P,
+    input_ref: P,
+    output_bam: P,
+    output_ref: P,
+    chr: String,
+    start: u64,
+    end: u64,
+) -> Result<()> {
+    let mut fasta_reader = fasta::IndexedReader::from_file(&input_ref)?;
+    fasta_reader.fetch(&chr, start - 1, end - 1)?;
+    let mut reference = Vec::new();
+    fasta_reader.read(&mut reference)?;
+    let mut rng = rand::thread_rng();
+    let alphabet = [b'A', b'C', b'G', b'T'];
+
+    //Build artificial reference
+    let mut artificial_reference = Vec::new();
+    add_random_bases(end - start, &mut artificial_reference, &mut rng, &alphabet)?;
+    let mut fa_writer = fasta::Writer::to_file(output_ref)?;
+    let ref_id = Uuid::new_v4().to_hyphenated().to_string();
+    fa_writer.write(&ref_id, None, &artificial_reference)?;
+
+    let mut bam_reader = bam::IndexedReader::from_path(bam)?;
+    bam_reader.fetch((chr.as_bytes(), start - 1, end))?;
+
+    let mut header = bam::Header::new();
+    header.push_record(
+        bam::header::HeaderRecord::new(b"SQ")
+            .push_tag(b"SN", &ref_id)
+            .push_tag(b"LN", &(end - start)),
+    );
+    let mut bam_writer = bam::Writer::from_path(output_bam, &header, bam::Format::Bam)?;
+    for result in bam_reader.records() {
+        let mut record = result?;
+        if (record.pos() >= (start - 1) as i64)
+            && (record.cigar().end_pos() < (end - 1) as i64)
+            && (record.mtid() == record.tid())
+            && (record.mpos() + 1 >= (start as i64))
+            && (record.mpos() + 1 < (end as i64))
+        //TODO Check if mate is fully in interval
+        {
+            record.cache_cigar();
+            //Check if mate record end within region
+            let artificial_seq = if record.is_unmapped() {
+                let mut seq = Vec::new();
+                add_random_bases(record.seq_len() as u64, &mut seq, &mut rng, &alphabet)?;
+                seq
+            } else {
+                build_sequence(
+                    &reference,
+                    &artificial_reference,
+                    &record,
+                    (start - 1) as usize,
+                    &mut rng,
+                    &alphabet,
+                )?
+            };
+            let artificial_record = build_record(&record, &artificial_seq, (start - 1) as i64)?;
+            bam_writer.write(&artificial_record)?;
+        }
+    }
+    Ok(())
+}
+
+fn build_record(record: &bam::Record, artificial_seq: &[u8], offset: i64) -> Result<bam::Record> {
+    let mut artificial_record = bam::record::Record::new();
+    if let Ok(mate_cigar) = record.aux(b"MC") {
```

Comment:
For some reason this line returns `[E::bam_aux_get] Corrupted aux data for read (null)` on each call even if the tag exists.

Reply:
Is the MC tag important to preserve?

Reply:
Yes. When using the data as input for fgbio it complains about missing MC tags.